cmd/bootstrap/utils/key_generation.go

```diff
@@ -18,6 +23,75 @@ func GenerateMachineAccountKey(seed []byte) (crypto.PrivateKey, error) {
 	return keys[0], nil
 }
 
+// The unstaked nodes have special networking keys, in two aspects:
+// - they use crypto.ECDSASecp256k1 keys, not crypto.ECDSAP256 keys,
+// - they use only positive keys (in the sense that the elliptic curve point of their public key is positive)
+//
+// Thanks to various properties of the cryptographic algorithm and libp2p,
+// this affords us to not have to maintain a table of flow.NodeID -> NetworkPublicKey
+// for those numerous and ephemeral nodes.
+// It incurs a one-bit security reduction, which is deemed acceptable.
+
+// drawUnstakedKey draws a single positive ECDSASecp256k1 key, and returns an error otherwise.
+func drawUnstakedKey(seed []byte) (crypto.PrivateKey, error) {
+	key, err := crypto.GeneratePrivateKey(crypto.ECDSASecp256k1, seed)
+	if err != nil {
+		// this should not happen
+		return nil, err
+	} else if key.PublicKey().EncodeCompressed()[0] == 0x03 {
+		// negative key -> unsuitable
+		return nil, fmt.Errorf("Unsuitable negative key")
+	}
+	return key, nil
+}
+
+// GenerateUnstakedNetworkingKey draws ECDSASecp256k1 keys until finding a suitable one.
+// though this will return fast, this is not constant-time and will leak ~1 bit of information through its runtime
+func GenerateUnstakedNetworkingKey(seed []byte) (crypto.PrivateKey, error) {
+	hkdf := hkdf.New(func() gohash.Hash { return sha256.New() }, seed, nil, []byte("unstaked network"))
+
+	round_seed := make([]byte, len(seed))
+	if _, err := io.ReadFull(hkdf, round_seed); err != nil {
+		// the hkdf Reader should not fail
+		panic(err)
+	}
+	max_iterations := 20 // 1/(2^20) failure chance
+	iteration_count := 0
+
+	var err error
+	var foundKey crypto.PrivateKey
+
+	for foundKey, err = drawUnstakedKey(round_seed); iteration_count < max_iterations && err != nil; foundKey, err = drawUnstakedKey(round_seed) {
+		if _, err := io.ReadFull(hkdf, round_seed); err != nil {
+			// the hkdf Reader should not fail
+			panic(err)
+		}
+		iteration_count++
+	}
+
+	if err != nil || iteration_count == max_iterations {
+		return nil, err
+	}
+	return foundKey, nil
+}
```

Comment:
```suggestion
func GenerateUnstakedNetworkingKey(seed []byte) (key crypto.PrivateKey, err error) {
	hkdf := hkdf.New(func() gohash.Hash { return sha256.New() }, seed, nil, []byte("unstaked network"))

	round_seed := make([]byte, len(seed))
	max_iterations := 20 // 1/(2^20) failure chance

	for i := 0; i < max_iterations; i++ {
		if _, err = io.ReadFull(hkdf, round_seed); err != nil {
			// the hkdf Reader should not fail
			panic(err)
		}
		if key, err = drawUnstakedKey(round_seed); err == nil {
			return
		}
	}

	return
}
```

Personally, I find this cleaner and easier to read.

Reply:
Yup, thanks!